kong/core/balancer.lua

```diff
@@ -1,41 +1,224 @@
+  -- re-created the balancer
+  balancers[upstream.name] = nil
+  
+  local b, err = ring_balancer.new({
+      wheelsize = upstream.slots,
+      order = upstream.orderlist,
+      dns = dns_client,
+    })
+  if not b then return b, err end
+  
+  -- NOTE: we're inserting a foreign entity in the balancer, to keep track of
+  -- target-history changes!
+  b.__targets_history = {} 
+  balancers[upstream.name] = b
 
-local first_try_balancer = function(target)
+  return upstream
 end
 
-local retry_balancer = function(target)
+-- finds and returns an upstream entity. This functions covers
+-- caching, invalidation, db access, et al.
+-- @return upstream table, or `false` if not found, or nil+error
+local function get_upstream(upstream_name)
+  local upstreams_dict, err = cache.get_or_set(cache.upstreams_dict_key(), 
+                                        nil, load_upstreams_dict_into_memory)
+  if err then
+    return nil, err
+  end
+
+  local upstream_id = upstreams_dict[upstream_name]
+  if not upstream_id then return false end -- no upstream by this name
+  
+  return cache.get_or_set(cache.upstream_key(upstream_id), nil, load_upstream_into_memory, upstream_id)
 end
 
-local first_try_dns = function(target)
-  local ip, port = toip(target.upstream.host, target.upstream.port, false)
-  if not ip then
-    return nil, port
+-- loads the target history for an upstream
+-- @param upstream_id Upstream uuid for which to load the target history
+local function load_targets_into_memory(upstream_id)
+  ngx_log(ngx_DEBUG, "fetching targets for upstream: ",tostring(upstream_id))
+  local target_history, err = singletons.dao.targets:find_all {upstream_id = upstream_id}
+  if err then
+    return nil, err
   end
-  target.ip = ip
-  target.port = port
-  return true
+  
+  -- some raw data updates
+  for _, target in ipairs(target_history) do
+    -- split `target` field into `name` and `port`
+    local port
+    target.name, port = string.match(target.target, "^(.-):(%d+)$")
+    target.port = tonumber(port)
+    -- need exact order, so order by created time and uuid
+    target.order = target.created_at..":"..target.id
+  end
+  
+  -- order by time
+  table.sort(target_history, function(a,b) return a.order<b.order end)
+--for i, t in ipairs(target_history) do
+--  ngx_log(ngx_DEBUG, i, t.order)
+--end
+  return target_history
 end
 
-local retry_dns = function(target)
-  local ip, port = toip(target.upstream.host, target.upstream.port, true)
-  if type(ip) ~= "string" then
-    return nil, port
+-- applies the history of lb transactions from index `start` forward
+-- @param rb ring-balancer object
+-- @param history list of targets/transactions to be applied
+-- @param start the index where to start in the `history` parameter
+-- @return true
+local function apply_history(rb, history, start)
+  
+  for i = start, #history do 
+    local target = history[i]
+    if target.weight > 0 then
```

Comment:
another thought here was that it might be reusable to use negative values for a special meaning, eg. `-1 == unhealthy` in case we might need to propagate that over the cluster.